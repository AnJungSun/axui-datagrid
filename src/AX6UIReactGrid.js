import React from 'react';
import UTIL from './AX6UIReactGrid-util';

//~~~~~


//~~~~~
class AX6UIReactGrid extends React.Component {
  constructor(props) {
    super(props);


    let stateForData = UTIL.getStateForData(this.props.data);
    // todo getStateForData 에서 컨버팅 할 데이터들 다시 확인

    // state init
    this.state = {
      dataOfList: stateForData.dataOfList,
      dataOfPage: stateForData.dataOfPage,
      proxyList: stateForData.proxyList,
      deletedList: [],

      bodyTrHeight: 0, // 한줄의 높이
      scrollContentWidth: 0, // 스크롤 될 내용물의 너비 (스크롤 될 내용물 : panel['body-scroll'] 안에 컬럼이 있는)
      scrollContentHeight: 0, // 스크롤 된 내용물의 높이
      scrollTimer: null,
      columns: [],
      colGroup: [],
      footSumColumns: [],
      bodyGrouping: [],
      sortInfo: {},
      focused: false,
      focusedColumn: {}, // 그리드 바디의 포커스된 셀 정보
      selectedColumn: {}, // 그리드 바디의 선택된 셀 정보
      isInlineEditing: false,
      inlineEditing: {},
      listIndexMap: {}, // tree데이터 사용시 데이터 인덱싱 맵
      headerTable: {},
      leftHeaderData: {},
      headerData: {},
      rightHeaderData: {},
      bodyRowTable: {},
      leftBodyRowData: {},
      bodyRowData: {},
      rightBodyRowData: {},
      bodyRowMap: {},
      bodyGroupingTable: {},
      leftBodyGroupingData: {},
      bodyGroupingData: {},
      rightBodyGroupingData: {},
      bodyGroupingMap: {},
      footSumTable: {}, // footSum의 출력레이아웃
      leftFootSumData: {}, // frozenColumnIndex 를 기준으로 나누어진 출력 레이아웃 왼쪽
      footSumData: {}, // frozenColumnIndex 를 기준으로 나누어진 출력 레이아웃 오른쪽
      needToPaintSum: true, // 데이터 셋이 변경되어 summary 변경 필요여부
    };
    this.dom = {
      container_hidden: null,
      container_header: null,
      container_body: null,
      container_page: null
    };

  }

  componentDidMount() {
    // setData.call(this);

    // this.dom.container_hidden.style.display = "block";
  }

  componentWillUnmount() {

  }

  componentWillReceiveProps(newProps) {

    // 새로운 array가 속성값으로 왔을경우
    // console.log("this.props.data === newProps.data", this.props.data === newProps.data);
    if (this.props.data !== newProps.data) {
      // 데이터가 변경된 경우
      let stateForData = UTIL.getStateForData(newProps.data);
      this.setState({
        dataOfList: stateForData.dataOfList,
        dataOfPage: stateForData.dataOfPage,
        proxyList: stateForData.proxyList,
        deletedList: []
      });
    }
  }

  render() {
    return (
      <div data-ax6ui-grid>
        <div data-ax6grid-container="root" style={{height: this.props.height}}>
          <div data-ax6grid-container="hidden" ref={ref => this.dom.container_hidden = ref}>
            <textarea data-ax6grid-form="clipboard" ref={ref => this.dom.clipboard = ref} />
          </div>
          <div data-ax6grid-container="header" ref={ref => this.dom.container_header = ref}>
            <div data-ax6grid-panel="aside-header"></div>
            <div data-ax6grid-panel="left-header"></div>
            <div data-ax6grid-panel="header">
              <div data-ax6grid-panel-scroll="header"></div>
            </div>
            <div data-ax6grid-panel="right-header"></div>
          </div>
          <div data-ax6grid-container="body" ref={ref => this.dom.container_body = ref}>
            <div data-ax6grid-panel="top-aside-body"></div>
            <div data-ax6grid-panel="top-left-body"></div>
            <div data-ax6grid-panel="top-body">
              <div data-ax6grid-panel-scroll="top-body"></div>
            </div>
            <div data-ax6grid-panel="top-right-body"></div>
            <div data-ax6grid-panel="aside-body">
              <div data-ax6grid-panel-scroll="aside-body"></div>
            </div>
            <div data-ax6grid-panel="left-body">
              <div data-ax6grid-panel-scroll="left-body"></div>
            </div>
            <div data-ax6grid-panel="body">
              <div data-ax6grid-panel-scroll="body">

                {this.state.dataOfList.length}

              </div>
            </div>
            <div data-ax6grid-panel="right-body">
              <div data-ax6grid-panel-scroll="right-body"></div>
            </div>
            <div data-ax6grid-panel="bottom-aside-body"></div>
            <div data-ax6grid-panel="bottom-left-body"></div>
            <div data-ax6grid-panel="bottom-body">
              <div data-ax6grid-panel-scroll="bottom-body"></div>
            </div>
            <div data-ax6grid-panel="bottom-right-body"></div>
          </div>
          <div data-ax6grid-container="page" ref={ref => this.dom.container_page = ref}>
            <div data-ax6grid-page="holder">
              <div data-ax6grid-page="navigation"></div>
              <div data-ax6grid-page="status"></div>
            </div>
          </div>
          <div data-ax6grid-container="scroller">
            <div data-ax6grid-scroller="vertical">
              <div data-ax6grid-scroller="vertical-bar"></div>
            </div>
            <div data-ax6grid-scroller="horizontal">
              <div data-ax6grid-scroller="horizontal-bar"></div>
            </div>
            <div data-ax6grid-scroller="corner"></div>
          </div>
          <div data-ax6grid-resizer="vertical"></div>
          <div data-ax6grid-resizer="horizontal"></div>
        </div>
      </div>
    );
  }
}

AX6UIReactGrid.defaultProps = {
  data: [],
  height: "300px",
  theme: 'default',
  animateTime: 250,
  debounceTime: 250,
  appendDebouncer: null,
  appendDebounceTimes: 0,
  appendProgressIcon: '...',
  appendProgress: false,

  // 틀고정 속성
  frozenColumnIndex: 0,
  frozenRowIndex: 0,
  showLineNumber: false,
  showRowSelector: false,
  multipleSelect: true,
  virtualScrollY: true,
  virtualScrollX: true,

  // 스크롤될 때 body 페인팅 딜레이를 주어 성능이 좋은 않은 브라우저에서 반응을 빠르게 할 때 사용하는 옵션들
  virtualScrollYCountMargin: 0,
  virtualScrollAccelerated: true,
  virtualScrollAcceleratedDelayTime: 10,

  columnMinWidth: 100,
  lineNumberColumnWidth: 30,
  rowSelectorColumnWidth: 26,
  sortable: undefined,
  remoteSort: false,

  header: {
    display: true,
    align: false,
    columnHeight: 26,
    columnPadding: 3,
    columnBorderWidth: 1,
    selector: true
  },
  body: {
    align: false,
    columnHeight: 26,
    columnPadding: 3,
    columnBorderWidth: 1,
    grouping: false,
    mergeCells: false
  },
  rightSum: false,
  footSum: false,
  page: {
    height: 25,
    display: true,
    statusDisplay: true,
    navigationItemCount: 5
  },
  scroller: {
    size: 15,
    barMinSize: 15,
    trackPadding: 4
  },
  columnKeys: {
    selected: '__selected__',
    modified: '__modified__',
    deleted: '__deleted__',
    disableSelection: '__disable_selection__'
  },
  tree: {
    use: false,
    hashDigit: 8,
    indentWidth: 10,
    arrowWidth: 15,
    iconWidth: 18,
    icons: {
      openedArrow: '▾',
      collapsedArrow: '▸',
      groupIcon: '⊚',
      collapsedGroupIcon: '⊚',
      itemIcon: '⊙'
    },
    columnKeys: {
      parentKey: "pid",
      selfKey: "id",
      collapse: "collapse",
      hidden: "hidden",
      parentHash: "__hp__",
      selfHash: "__hs__",
      children: "__children__",
      depth: "__depth__",
    }
  }
};

export default AX6UIReactGrid