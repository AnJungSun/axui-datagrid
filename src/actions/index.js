import {isArray, isObject} from "underscore";
import { propsConverterForData } from "../_inc/utils";

/* 그리드 초기화 :  */
export const INIT = 'INIT';

export function init(props) {
  let Obj_data = propsConverterForData(props.data);
  return {
    type: INIT,
    height: props.height,
    columns: props.columns,
    receivedList: Obj_data.receivedList,
    page: Obj_data.page,
    options: props.options
  }
}

export const DID_MOUNT = 'DID_MOUNT';
export function didMount(props, containerDOM) {
  return {
    type: DID_MOUNT,
    containerDOM: containerDOM
  }
}

/* 그리드 데이터 변경 */
export const SET_DATA= 'SET_DATA';
export function setData(data) {
  let Obj_data = propsConverterForData(data);

  return {
    type: SET_DATA,
    receivedList: Obj_data.receivedList,
    page: Obj_data.page,
  }
}

/* 그리드 스크롤 업데이트 */
export const UPDATE_SCROLL = 'UPDATE_SCROLL';
export function updateScroll(scrollLeft, scrollTop) {
  return {
    type: UPDATE_SCROLL,
    scrollLeft,
    scrollTop,
  }
}