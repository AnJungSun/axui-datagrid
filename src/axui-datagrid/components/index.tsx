export { default as DataGridBody } from './DataGridBody';
export { default as DataGridBodyCell } from './DataGridBodyCell';
export { default as DataGridBodyPanel } from './DataGridBodyPanel';
export { default as DataGridColumnFilter } from './DataGridColumnFilter';
export {
  default as DataGridColumnFilterOption,
} from './DataGridColumnFilterOption';
export { default as DataGridHeader } from './DataGridHeader';
export { default as DataGridHeaderCell } from './DataGridHeaderCell';
export { default as DataGridHeaderPanel } from './DataGridHeaderPanel';
export { default as DataGridPage } from './DataGridPage';
export { default as DataGridRoot } from './DataGridRoot';
export { default as DataGridScroll } from './DataGridScroll';
export { default as DataGridSelector } from './DataGridSelector';
