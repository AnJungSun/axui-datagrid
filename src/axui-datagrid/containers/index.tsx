export { default as AXUIDatagrid } from './Datagrid';
