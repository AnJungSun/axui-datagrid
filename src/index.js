import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import gridReducers from './reducers';
import GridContainer from './container/GridContainer';

//~~~~~
const store = createStore(gridReducers, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__());

//~~~~~
class AX6UIReactGrid extends React.Component {
  render() {
    return (
      <Provider store={store}>
        <GridContainer {...this.props}/>
      </Provider>
    );
  }
}

export default AX6UIReactGrid