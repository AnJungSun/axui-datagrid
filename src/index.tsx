import * as React from 'react';
import {connect, Provider} from 'react-redux';
import {store} from './store';
import * as act from './actions';
import {GridRoot} from './component/GridRoot';

interface iProps {
  gridCSS: object;
  height: string;
  style: object;
  columns: any;
  data: any;
  options: object;
  thisCallback: object;
}

export const GridRootConnected = connect(
  (state: any) => {
    return {
      store_receivedList: state.get('receivedList'),
      store_deletedList: state.get('deletedList'),
      store_list: state.get('list'),
      store_page: state.get('page'),
      store_sortInfo: state.get('sortInfo')
    };
  },
  (dispatch: any) => ({
    init: (props, options) => dispatch(act.init(props, options)),
    setData: (data, options) => dispatch(act.setData(data, options))
  })
)(GridRoot);

class AXDatagrid extends React.Component<iProps, {}> {
  render() {
    return (
      <Provider store={store}>
        <GridRootConnected {...this.props} />
      </Provider>
    );
  }
}

export {
  AXDatagrid
};
