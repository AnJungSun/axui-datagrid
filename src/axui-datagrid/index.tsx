import { DataGrid } from './containers';
import { typeStore } from './stores';

export { typeStore, DataGrid };
