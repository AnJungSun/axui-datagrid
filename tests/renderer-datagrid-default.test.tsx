import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { DataGrid } from '../src/axui-datagrid';

it('renders DataGrid default', () => {
  let gridData: any[] = [];

  const state = {
    height: 300,
    columns: [
      {
        key: 'a',
        label: '필드A',
        width: 80,
        align: 'center',
      },
      { key: 'b', label: '필드B', align: 'center' },
      { key: 'c', label: '필드C', align: 'center' },
      { key: 'price', label: '단가', formatter: 'money', align: 'right' },
      { key: 'amount', label: '수량', formatter: 'money', align: 'right' },
      { key: 'cost', label: '금액', align: 'right', formatter: 'money' },
      { key: 'saleDt', label: '판매일자', align: 'center' },
      { key: 'customer', label: '고객명', align: 'center' },
      { key: 'saleType', label: '판매타입', align: 'center' },
    ],
    data: gridData,
    options: {
      lineNumberColumnWidth: 60,
      showLineNumber: true,
      showRowSelector: false,
    },
  };

  const tree = renderer
    .create(
      <DataGrid
        columns={state.columns}
        height={state.height}
        data={state.data}
        options={state.options}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
