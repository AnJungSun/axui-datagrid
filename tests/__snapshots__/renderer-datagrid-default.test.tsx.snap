// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DataGrid default 1`] = `
<div
  className="axui-datagrid"
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchStartCapture={[Function]}
  onWheel={[Function]}
  style={
    Object {
      "height": 300,
    }
  }
  tabIndex={-1}
>
  <div
    className="axui-datagrid-clip-board"
  >
    <textarea />
  </div>
  <div
    className="axui-datagrid-header"
    style={
      Object {
        "height": 24,
      }
    }
  >
    <div
      data-panel="aside-header"
      style={
        Object {
          "height": 24,
          "left": 0,
          "width": 60,
        }
      }
    >
      <table
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <colgroup>
          <col
            style={
              Object {
                "width": "60px",
              }
            }
          />
          <col />
        </colgroup>
        <tbody>
          <tr
            className=""
          >
            <td
              className="axui-datagrid-header-column axui-datagrid-header-corner"
              colSpan={1}
              onClick={[Function]}
              rowSpan={1}
              style={
                Object {
                  "height": 23,
                  "minHeight": "1px",
                }
              }
            >
              <span
                data-align="center"
                data-span={true}
                style={
                  Object {
                    "height": "23px",
                    "lineHeight": "17px",
                  }
                }
              >
                 
              </span>
            </td>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
    <div
      data-scroll-container="header-scroll-container"
      style={
        Object {
          "height": 24,
          "left": 60,
          "width": -60,
        }
      }
    >
      <div
        data-panel="header-scroll"
        style={
          Object {
            "height": 24,
            "left": 0,
          }
        }
      >
        <table
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "80px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col />
          </colgroup>
          <tbody>
            <tr
              className=""
            >
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="center"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  필드A
                </span>
                <span
                  data-filter="true"
                  data-filter-index={0}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="center"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  필드B
                </span>
                <span
                  data-filter="true"
                  data-filter-index={1}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="center"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  필드C
                </span>
                <span
                  data-filter="true"
                  data-filter-index={2}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="right"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  단가
                </span>
                <span
                  data-filter="true"
                  data-filter-index={3}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="right"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  수량
                </span>
                <span
                  data-filter="true"
                  data-filter-index={4}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="right"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  금액
                </span>
                <span
                  data-filter="true"
                  data-filter-index={5}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="center"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  판매일자
                </span>
                <span
                  data-filter="true"
                  data-filter-index={6}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="center"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  고객명
                </span>
                <span
                  data-filter="true"
                  data-filter-index={7}
                />
              </td>
              <td
                className="axui-datagrid-header-column"
                colSpan={1}
                onClick={[Function]}
                rowSpan={1}
                style={
                  Object {
                    "height": 23,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-align="center"
                  data-span={true}
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                    }
                  }
                >
                  판매타입
                </span>
                <span
                  data-filter="true"
                  data-filter-index={8}
                />
              </td>
              <td />
            </tr>
          </tbody>
        </table>
        <div
          data-column-resizer={0}
          data-left={80}
          data-prev-left={0}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "78px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={1}
          data-left={180}
          data-prev-left={80}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "178px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={2}
          data-left={280}
          data-prev-left={180}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "278px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={3}
          data-left={380}
          data-prev-left={280}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "378px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={4}
          data-left={480}
          data-prev-left={380}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "478px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={5}
          data-left={580}
          data-prev-left={480}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "578px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={6}
          data-left={680}
          data-prev-left={580}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "678px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={7}
          data-left={780}
          data-prev-left={680}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "778px",
              "width": 4,
            }
          }
        />
        <div
          data-column-resizer={8}
          data-left={880}
          data-prev-left={780}
          onMouseDown={[Function]}
          style={
            Object {
              "height": "23px",
              "left": "878px",
              "width": 4,
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="axui-datagrid-body"
    onMouseDown={[Function]}
    style={
      Object {
        "height": 256,
      }
    }
  >
    <div
      data-scroll-container="aside-body-scroll-container"
      style={
        Object {
          "height": 256,
          "left": 0,
          "top": 0,
          "width": 60,
        }
      }
    >
      <div
        data-panel="aside-body-scroll"
        style={
          Object {
            "left": 0,
            "paddingLeft": 0,
            "paddingTop": 0,
            "top": 0,
          }
        }
      >
        <table
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "60px",
                }
              }
            />
            <col />
          </colgroup>
          <tbody />
        </table>
      </div>
    </div>
    <div
      data-scroll-container="body-scroll-container"
      style={
        Object {
          "height": 256,
          "left": 60,
          "top": 0,
          "width": -60,
        }
      }
    >
      <div
        data-panel="body-scroll"
        style={
          Object {
            "left": 0,
            "paddingLeft": 0,
            "paddingTop": 0,
            "top": 0,
          }
        }
      >
        <table
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "80px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col />
          </colgroup>
          <tbody />
        </table>
      </div>
    </div>
    <div
      data-scroll-container="bottom-aside-body-scroll-container"
      style={
        Object {
          "height": 0,
          "left": 0,
          "top": 256,
          "width": 60,
        }
      }
    >
      <div
        data-panel="bottom-aside-body-scroll"
        style={
          Object {
            "left": 0,
            "paddingLeft": 0,
            "paddingTop": 0,
            "top": 0,
          }
        }
      >
        <table
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "60px",
                }
              }
            />
            <col />
          </colgroup>
          <tbody>
            <tr
              className=""
            >
              <td
                className="axui-datagrid-line-number"
                colSpan={1}
                rowSpan={1}
                style={
                  Object {
                    "height": 24,
                    "minHeight": "1px",
                  }
                }
              >
                <span
                  data-pos={-1}
                  data-span="lineNumber"
                  style={
                    Object {
                      "height": "23px",
                      "lineHeight": "17px",
                      "textAlign": "center",
                    }
                  }
                >
                  
                </span>
              </td>
              <td />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      data-scroll-container="bottom-body-scroll-container"
      style={
        Object {
          "height": 0,
          "left": 60,
          "top": 256,
          "width": -60,
        }
      }
    >
      <div
        data-panel="bottom-body-scroll"
        style={
          Object {
            "left": 0,
            "paddingLeft": 0,
            "paddingTop": 0,
            "top": 0,
          }
        }
      >
        <table
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "80px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "100px",
                }
              }
            />
            <col />
          </colgroup>
          <tbody>
            <tr
              className=""
            >
              <td />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    className="axui-datagrid-page"
    style={
      Object {
        "height": 20,
      }
    }
  >
    <div
      className="axui-datagrid-page-buttons"
      style={
        Object {
          "width": 150,
        }
      }
    >
      <button
        onClick={[Function]}
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <div
          className="datagridIcon-first"
          data-button-svg={true}
        />
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <div
          className="datagridIcon-prev"
          data-button-svg={true}
        />
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <div
          className="datagridIcon-back"
          data-button-svg={true}
        />
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <div
          className="datagridIcon-play"
          data-button-svg={true}
        />
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <div
          className="datagridIcon-next"
          data-button-svg={true}
        />
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <div
          className="datagridIcon-last"
          data-button-svg={true}
        />
      </button>
    </div>
  </div>
</div>
`;
